gpbft/gpbft.go

```diff
@@ -579,27 +579,32 @@ func (i *instance) tryConverge() error {
 	if i.phase != CONVERGE_PHASE {
 		return fmt.Errorf("unexpected phase %s, expected %s", i.phase, CONVERGE_PHASE)
 	}
+	// The CONVERGE phase timeout doesn't wait to hear from >⅔ of power.
 	timeoutExpired := atOrAfter(i.host.Time(), i.phaseTimeout)
 	if !timeoutExpired {
 		return nil
 	}
 
-	i.value = i.roundState(i.round).converged.findMaxTicketProposal(i.powerTable)
-	if i.value.IsZero() {
+	possibleDecisionLastRound := !i.roundState(i.round - 1).committed.HasStrongQuorumFor("")
+	winner := i.roundState(i.round).converged.findMaxTicketProposal(i.powerTable)
+	if winner.Chain.IsZero() {
 		return fmt.Errorf("no values at CONVERGE")
 	}
-	if i.isAcceptable(i.value) {
-		// Sway to proposal if the value is acceptable.
-		if !i.proposal.Eq(i.value) {
-			i.proposal = i.value
-			i.log("adopting proposal %s after converge", &i.proposal)
-		}
-	} else {
-		// Vote for not deciding in this round
-		i.value = ECChain{}
+	// If the winner is not a candidate but it could possibly have been decided by another participant
+	// in the last round, consider it a candidate.
+	if !i.isCandidate(winner.Chain) && winner.Justification.Vote.Step == PREPARE_PHASE && possibleDecisionLastRound {
+		i.log("⚠️ swaying from %s to %s by CONVERGE", &i.proposal, &winner.Chain)
+		i.candidates = append(i.candidates, winner.Chain)
 	}
-	i.beginPrepare()
+	if i.isCandidate(winner.Chain) {
+		i.proposal = winner.Chain
+		i.log("adopting proposal %s after converge", &winner.Chain)
+	}
+	// Else preserve own proposal
+	// XXX spec says to loop to next lowest ticket, rather than fall back to own proposal.
```

Comment:
I'll delete this comment before merging, either because we decide it's ok or because I've added a task in #170 to do the iteration

Reply:
Thanks. Consider simplifying the FIP psuedocode too, and just noting the loop as a potential optimisation.